examples/long_running_subscriber/main.go

```diff
@@ -0,0 +1,364 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/sailhouse/sdk-go/sailhouse"
+)
+
+// Stats for tracking processing
+type ProcessingStats struct {
+	eventsProcessed  atomic.Int64
+	processingErrors atomic.Int64
+	handlerErrors    atomic.Int64
+}
+
+func (s *ProcessingStats) String() string {
+	return fmt.Sprintf("Events: %d, Processing Errors: %d, Handler Errors: %d",
+		s.eventsProcessed.Load(),
+		s.processingErrors.Load(),
+		s.handlerErrors.Load())
+}
+
+func main() {
+	// Parse command line flags
+	token := flag.String("token", "", "Sailhouse API token")
+	processors := flag.Int("processors", 2, "Number of processors per subscription")
+	flag.Parse()
+
+	if *token == "" {
+		log.Fatal("Usage: go run main.go -token=<token> [-processors=<count>]")
+	}
+
+	// Create context that cancels on interrupt
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Handle graceful shutdown
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	// Create Sailhouse client
+	client := sailhouse.NewSailhouseClient(*token)
+
+	// Create processing stats
+	stats := &ProcessingStats{}
+
+	// Create subscriber with custom options
+	options := &sailhouse.SubscriberOptions{
+		PerSubscriptionProcessors: *processors,
+		PollInterval:              500 * time.Millisecond,
+		MaxRetries:                3,
+		RetryDelay:                2 * time.Second,
+		ErrorHandler: func(err error) {
+			stats.processingErrors.Add(1)
+			log.Printf("Processing error: %v", err)
+		},
+	}
+
+	subscriber := client.NewSubscriber(options)
+
+	// Register subscription handlers
+	registerHandlers(subscriber, stats)
+
+	// Start the subscriber
+	fmt.Printf("Starting subscriber with %d processors per subscription...\n", *processors)
+	err := subscriber.Start(ctx)
+	if err != nil {
+		log.Fatalf("Failed to start subscriber: %v", err)
+	}
+
+	fmt.Println("Subscriber started successfully!")
+	fmt.Println("Registered subscriptions:")
+	for _, sub := range subscriber.GetSubscriptions() {
+		fmt.Printf("  - %s/%s\n", sub.Topic, sub.Subscription)
+	}
+	fmt.Println()
+
+	// Start stats reporting
+	go reportStats(ctx, stats)
+
+	// Wait for shutdown signal
+	select {
+	case sig := <-sigChan:
+		fmt.Printf("\nReceived signal %v, shutting down gracefully...\n", sig)
+	case <-ctx.Done():
+		fmt.Println("\nContext cancelled, shutting down...")
+	}
+
+	// Graceful shutdown
+	fmt.Println("Stopping subscriber...")
+	subscriber.Stop()
+	fmt.Println("Subscriber stopped.")
+	fmt.Printf("Final stats: %s\n", stats)
+}
+
+func registerHandlers(subscriber *sailhouse.SailhouseSubscriber, stats *ProcessingStats) {
+	// Handler for user events
+	subscriber.Subscribe("user-events", "user-processor", func(ctx context.Context, event *sailhouse.Event) error {
```

Comment:
Each event type should be it's own topic imo?

Reply:
AI... Agreed